include/verilatedos.h

```diff
@@ -161,6 +161,13 @@
 # define VL_MT_SAFE
 #endif
 // Comment tag that function is threadsafe, only
+// when other theads doesn't change AST tree
+#if defined(__clang__)
+# define VL_MT_SAFE_STABLE_TREE __attribute__((annotate("MT_SAFE")))
```

Comment:
Just to make sure I understood your comment correctly: I should add ``VL_MT_SAFE_STABLE`` definition (as I can't see ``stable`` defined already) and add check that ``VL_MT_SAFE_STABLE`` doesn't call ``VL_MT_SAFE_STABLE_TREE``? If yes, when node should be annotated with ``VL_MT_SAFE_STABLE``?

Reply:
My comment was confusing, I meant do this

     # define VL_MT_SAFE_STABLE_TREE __attribute__((annotate("MT_SAFE_STABLE_TREE")))

Then update the checker that MT_SAFE_STABLE_TREE can call other MT_SAFE_STABLE_TREE or MT_SAFE or PURE (i.e. anything marked). But MT_SAFE cannot call MT_SAFE_STABLE_TREE as that is "less" safe. The idea is from a safeness standpoint we enforce

     MT_SAFE_STABLE_TREE < MT_SAFE < PURE

Reply:
The main goal of this PR is to make ``v3errorEnd`` safe to call from MT-safe context (to allow parallelization). I believe that this check (on ``v3errorEnd`` function) would mean that no part of parallel code can call ``v3errorEnd`` or whole speed-up from parallelization would be gone.

If I'm not wrong, the only part that makes ``v3errorEnd`` require stable tree are functions inside ``V3ERROR_NO_GLOBAL_`` ifndef (https://github.com/verilator/verilator/blob/master/src/V3Error.cpp#L279).
They are only called when error is FATAL and debug or dump tree flags are set. 

What to you think about not calling this function if we are running verilation with multiple threads and instead print some information message that it is only supported on single thread verilation? This would allow to annotate ``v3errorEnd`` with just ``MT_SAFE``.

Reply:
I think we definitely do want to dump when multithreaded.  I'm expecting at some point that will be the main operation mode and getting these dumps is critical to debug.

It makes sense that most algorithms will want to have routines that execute in parallel, and inside that parallel code they have UASSERTs.  We don't want to have to single thread those UASSERTs just in case they fire.  How in general will we be doing thread cleanup on fatals?  That answer might suggest how we handle v3fatal.

Reply:
Sure, so I think it would be the best to use ``Cooperative Interruption`` concept from C++20 to handle this (https://en.cppreference.com/w/cpp/thread/stop_source / https://www.modernescpp.com/index.php/cooperative-interruption-of-a-thread-in-c-20).

In this concept we can emit request that other threads can check (but this needs to be done manually) in appropriate place and if it is set, stop executing the job.

When we will encourage fatal error that requires dumping the tree, we would emit such signal, wait for other threads to stop execution (by entering function that checks for stop request, so generally any function that requires stable tree, or completes its job).
When all threads will complete its job, we would continue main thread and dump tree.

What do you think about this solution?

As this is mostly used by developers, maybe we can require C++20 for this feature?